### PR TITLE
[#25] Implemented dashboard page with preview mode and gridView mode.

### LINKED ISSUE #25: Design and implement dashboard page.
- Queries thumbnail images and first episode previews form the server.
- Get webtoon infos categorized by days.
- Able to convert preview mode to list view mode or vice verse.

### PR DESCRIPTION
**Pull request by harry2636**
Fixes #25 

- Queries thumbnail images and first episode previews form the server.
- Get webtoon infos categorized by days.
- Able to convert preview mode to gridView mode with button on the menu.
- Add sample webtoon list view page and read page.

### CODE REVIEW

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+ */
+public class PreviewModeFragment extends Fragment {
+  private static final int MAX_DOWNLOAD_SIZE_BYTES = 1024 * 1024 * 10; // 10MiB
+  private Thumbnail currentWebtoon = null;
+  private long lastClickTime = 0;
+
+  @Override
+  public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                           Bundle savedInstanceState) {
+    View view = inflater.inflate(R.layout.previewmode_fragment_main, container, false );
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java]: space before `)`?

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+          thumb.thumbLayout.setOnClickListener(onClickListener);
+          thumb.thumbLayout.setOnLongClickListener(onLongClickListener);
+
+          getActivity().runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+              leftLinear.addView(thumb);
+            }
+          });
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java:111]: unnecessary blank line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+          for (Element imgLink : wtViewer.children()) {
+            String imgUrl = imgLink.absUrl("src");
+
+            // Check whether imgURL is a valid image file
+            if (imgUrl == null || !imgUrl.endsWith(".jpg")) {
+              continue;
+            }
+
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java]: unnecessary blank line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      public void run() {
+        try {
+          Element wtPage;
+          wtPage = Jsoup.connect(firstEpisodeUrl)
+              .userAgent("Mozilla/5.0 (Windows NT 6.1; WOW64; rv:23.0)"
+                  + "Gecko/20100101 Firefox/23.0")
+              .get();
+          Element wtViewer = wtPage.getElementsByClass("wt_viewer").first();
+
+          for (Element imgLink : wtViewer.children()) {
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java:154]: I get `NullPointerException` here. It seems that we need to handle the case where the page does not have DOM element having class `wt_viewer`.
How about adding a catch statement?

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/ReadEpisodePage.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      BitmapFactory.Options options = new BitmapFactory.Options();
+      Bitmap bitmap = BitmapFactory.decodeResource(getResources(), res, options);
+      imageView.setImageBitmap(bitmap);
+      LinearLayout.LayoutParams layoutParams = new LinearLayout
+          .LayoutParams(MATCH_PARENT, WRAP_CONTENT);
+      imageView.setLayoutParams(layoutParams);
+      imageView.setAdjustViewBounds(true);
+      imageView.setScaleType(ImageView.ScaleType.FIT_START);
+
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/ReadEpisodePage.java]: unnecessary blank line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/ReadEpisodePage.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      readLinear.addView(imageView);
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/ReadEpisodePage.java]: UBL (unnecessary blank line)

File: Android/toonDra/app/src/main/res/layout/episode_read_page.xml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    </LinearLayout>
+  </ScrollView>
+  <ImageView
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      app:srcCompat="@drawable/eye_temp"
+      android:id="@+id/eye_screen_center"
+      android:layout_gravity="center"/>
+
+</FrameLayout>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/layout/episode_read_page.xml]: end of file with no extra line

File: Android/toonDra/app/src/main/res/layout/episodelist_fragment.xml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+              android:layout_width="match_parent"
+              android:layout_height="match_parent">
+
+
+  <ImageView
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:src="@drawable/toondra_listsample"
+      android:id="@+id/episodeListView"/>
+</LinearLayout>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/layout/episodelist_fragment.xml]: end of file with no blank line

File: Android/toonDra/app/src/main/res/layout/gridviewmode_fragment_main.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  android:layout_height="match_parent">
+
+  <GridView
+      android:id="@+id/fullGrid"
+      android:layout_width="match_parent"
+      android:layout_height="match_parent"
+      android:numColumns="3"
+      android:horizontalSpacing="1dp"
+      android:background="@color/borderColor"/>
+</LinearLayout>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/layout/gridviewmode_fragment_main.xml]: EOF with no extra line

File: Android/toonDra/app/src/main/res/layout/page_custom.xml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ImageView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/thumb_image"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:adjustViewBounds="true"
+/>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/layout/page_custom.xml]: EOF with no blank line

File: Android/toonDra/app/src/main/res/layout/previewmode_fragment_main.xml [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+          android:id="@+id/previewDescription"
+          android:layout_width="match_parent"
+          android:layout_height="500dp"
+          android:text="첫 화 미리보기를 할 웹툰을 눌러주세요"
+          android:textStyle="bold"
+          android:textSize="15sp"
+          android:gravity="center"/>
+    </LinearLayout>
+  </ScrollView>
+</LinearLayout>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/layout/previewmode_fragment_main.xml]: EOF with no extra line

File: Android/toonDra/app/src/main/res/layout/thumbnail_custom.xml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      android:id="@+id/thumb_text"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:textAppearance="?android:attr/textAppearanceSmall"
+      android:textColor="@android:color/black"
+      android:gravity="center"
+      android:layout_gravity="center_horizontal"
+      android:textStyle="bold"
+      android:background="@color/waitButtonColor"/>
+</LinearLayout>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/layout/thumbnail_custom.xml]: EOF with no extra line

File: Android/toonDra/app/src/main/res/menu/activity_main_drawer.xml [hunk 13]
```diff
@@ -33,4 +33,4 @@
         </menu>
     </item>
 
-</menu>
+</menu>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/menu/activity_main_drawer.xml]: EOF with no extra line

File: Android/toonDra/app/src/main/res/menu/main.xml [hunk 14]
```diff
@@ -1,9 +1,9 @@
...
-        android:orderInCategory="100"
-        android:title="@string/action_settings"
-        app:showAsAction="never" />
-</menu>
+    <!-- Change fragment -->
+    <item android:id="@+id/change_fragment"
+          android:icon="@drawable/gridview_menu"
+          android:title="change_fragment"
+          app:showAsAction="always"/>
+</menu>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/menu/main.xml]: EOF with no extra line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/App.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    float scale = getResources().getDisplayMetrics().density;
+    return (int) (dp * scale + 0.5f);
+  }
+
+  @Override
+  protected void attachBaseContext(Context base) {
+    super.attachBaseContext(base);
+    MultiDex.install(getBaseContext());
+  }
+}
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/App.java]: EOF with no extra line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/EpisodeListPage.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+    imageView.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        Intent intent = new Intent(getApplicationContext(), ReadEpisodePage.class);
+        startActivity(intent);
+      }
+    });
+  }
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/EpisodeListPage.java]: unnecessary blank line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/GridViewModeFragment.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+  public void addThumbnailsToGrid(Day day) {
+    GridView gridView = (GridView) getView().findViewById(R.id.fullGrid);
+    gridView.setAdapter(null);
+
+    NaverToonInfo[] webtoons = ((MainActivity)getActivity()).getCurrentWebtoonsByDay(day);
+    ThumbnailAdapter adapter = new ThumbnailAdapter(getActivity(), webtoons);
+    gridView.setAdapter(adapter);
+  }
+}
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/GridViewModeFragment.java]: unnecessary blank line

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/MainActivity.java [hunk 18]
```diff
@@ -115,4 +136,251 @@ public boolean onNavigationItemSelected(MenuItem item) {
...
+    FragmentManager fragmentManager = getFragmentManager();
+    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
+    fragmentTransaction.add(R.id.container, previewModeFragment);
+    fragmentTransaction.add(R.id.container, gridViewModeFragment);
+    fragmentTransaction.hide(gridViewModeFragment);
+    fragmentTransaction.commit();
+  }
 }
+
+
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/MainActivity.java]: unnecessary blank lines

**Review by dansuh17 — COMMENTED** [review]
review by me

**Comment by dansuh17** [comment]
Also, I get OoM (Out of Memory) Exception. I presume that this is due to large image file sizes. We have two options : 1. increase the heap size of JVM, 2. resize image files [see here](http://stackoverflow.com/questions/10413659/how-to-resize-image-in-android).

File: Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java (hunk 4 above)
Comment (harry2636) [reply Android/toonDra/app/src/main/java/edu/kaist/mskers/toondra/PreviewModeFragment.java:154]: I think NullPointerException only occurs when the user accesses the 19 years old webtoon. (It is blocked due to permission problem.) I will put alert message for this.

**Comment by harry2636** [comment]
@deNsuh Can you explain the situation where the OOM occurs? Does it occurs in read episode page?

**Comment by dansuh17** [comment]
@harry2636 Yes it occurs within the read episode. It rarely happens immediately after entering the view, but while scrolling.

**Review by heewoonam — CHANGES_REQUESTED** [review]
Null pointer exception when there is no Internet connection.

**Comment by harry2636** [comment]
@deNsuh @heewoonam 
I fixed all null pointer exception cases.
For the read page OOM, it occurs because the current read page uses "res/drawable" files. After I implement the read page with the remote server, the OOM bug will be automatically fixed.
I just commented the sample read page parts.

File: Android/toonDra/app/src/main/res/drawable/thumb_back.xml [hunk 19]
```diff
@@ -0,0 +1,4 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android" android:shape="rectangle" >
+  <solid android:color="#ffffff" />
+  <stroke android:width="0.3dip" android:color="#999999"/>
+</shape>
```
Comment (dansuh17) [inline Android/toonDra/app/src/main/res/drawable/thumb_back.xml]: one last empty line before EOF 😉 
LGTM except this.

**Comment by dansuh17** [comment]
LGTM, except for the one thing commented above :)

**Comment by harry2636** [comment]
All done, I will merge it, because @heewoonam said that he is busy so that we can merge the commit.

**Pull request merged**